Mathlib/Algebra/GroupPower/Lemmas.lean

```diff
@@ -2,35 +2,1316 @@
+    simp only [abs_nsmul, coe_nat_zsmul, Nat.abs_cast]
+  · obtain ⟨m, h⟩ := Int.eq_ofNat_of_zero_le (neg_nonneg.2 n0)
+    rw [← abs_neg, ← neg_zsmul, ← abs_neg n, h, coe_nat_zsmul, Nat.abs_cast, coe_nat_zsmul]
+    exact abs_nsmul m _
+#align abs_zsmul abs_zsmul
+
+theorem abs_add_eq_add_abs_le (hle : a ≤ b) : |a + b| = |a| + |b| ↔ 0 ≤ a ∧ 0 ≤ b ∨ a ≤ 0 ∧ b ≤ 0 :=
+  by
+  obtain a0 | a0 := le_or_lt 0 a <;> obtain b0 | b0 := le_or_lt 0 b
+  · simp [a0, b0, abs_of_nonneg, add_nonneg a0 b0]
+  · exact (lt_irrefl (0 : α) <| a0.trans_lt <| hle.trans_lt b0).elim
+  any_goals simp [a0.le, b0.le, abs_of_nonpos, add_nonpos, add_comm]
+  have : (|a + b| = -a + b ↔ b ≤ 0) ↔ (|a + b| = |a| + |b| ↔ 0 ≤ a ∧ 0 ≤ b ∨ a ≤ 0 ∧ b ≤ 0) := by
+    simp [a0, a0.le, a0.not_le, b0, abs_of_neg, abs_of_nonneg]
+  refine' this.mp ⟨fun h => _, fun h => by simp only [le_antisymm h b0, abs_of_neg a0, add_zero]⟩
+  obtain ab | ab := le_or_lt (a + b) 0
+  · refine' le_of_eq (eq_zero_of_neg_eq _)
+    rwa [abs_of_nonpos ab, neg_add_rev, add_comm, add_right_inj] at h
+  · refine' (lt_irrefl (0 : α) _).elim
+    rw [abs_of_pos ab, add_left_inj] at h
+    rwa [eq_zero_of_neg_eq h.symm] at a0
+#align abs_add_eq_add_abs_le abs_add_eq_add_abs_le
+
+theorem abs_add_eq_add_abs_iff (a b : α) : |a + b| = |a| + |b| ↔ 0 ≤ a ∧ 0 ≤ b ∨ a ≤ 0 ∧ b ≤ 0 := by
+  obtain ab | ab := le_total a b
+  · exact abs_add_eq_add_abs_le ab
+  · rw [add_comm a, add_comm (abs _), abs_add_eq_add_abs_le ab, and_comm, @and_comm (b ≤ 0)]
+#align abs_add_eq_add_abs_iff abs_add_eq_add_abs_iff
+
+end LinearOrderedAddCommGroup
+
+@[simp]
+theorem WithBot.coe_nsmul [AddMonoid A] (a : A) (n : ℕ) : (n • a : A) = n • (a : WithBot A) :=
+  AddMonoidHom.map_nsmul
+    { toFun := fun a : A => (a : WithBot A),
+      map_zero' := WithBot.coe_zero,
+      map_add' := WithBot.coe_add }
+    a n
+#align with_bot.coe_nsmul WithBot.coe_nsmul
+
+theorem nsmul_eq_mul' [NonAssocSemiring R] (a : R) (n : ℕ) : n • a = a * n := by
+  induction' n with n ih <;> [rw [zero_nsmul, Nat.cast_zero, mul_zero],
+    rw [succ_nsmul', ih, Nat.cast_succ, mul_add, mul_one]]
+#align nsmul_eq_mul' nsmul_eq_mul'ₓ
+-- typeclasses do not match up exactly.
+
+@[simp]
+theorem nsmul_eq_mul [NonAssocSemiring R] (n : ℕ) (a : R) : n • a = n * a := by
   rw [nsmul_eq_mul', (n.cast_commute a).eq]
+#align nsmul_eq_mul nsmul_eq_mulₓ
+-- typeclasses do not match up exactly.
+
+/-- Note that `AddCommMonoid.nat_smul_comm_class` requires stronger assumptions on `R`. -/
+instance NonUnitalNonAssocSemiring.nat_smul_comm_class [NonUnitalNonAssocSemiring R] :
+    SMulCommClass ℕ R R :=
+  ⟨fun n x y => by
+    induction' n with n ih
+    · simp [zero_nsmul]
+    · simp_rw [succ_nsmul, smul_eq_mul, mul_add, ← smul_eq_mul, ih]⟩
+#align
+  non_unital_non_assoc_semiring.nat_smul_comm_class NonUnitalNonAssocSemiring.nat_smul_comm_class
+
+/-- Note that `AddCommMonoid.nat_is_scalar_tower` requires stronger assumptions on `R`. -/
```

Comment:
That looks like a comment rather than a docstring, but I see it was in mathlib3 too...